x-pack/plugins/endpoint/server/routes/alerts/list/lib.ts

```diff
@@ -0,0 +1,233 @@
+  AlertDataWrapper,
+  AlertHits,
+  AlertResultList,
+  ESTotal,
+} from '../../../../common/types';
+import { EndpointAppContext } from '../../../types';
+import {
+  AlertListRequestParams,
+  AlertListRequestData,
+  AlertListRequest,
+  AlertListRequestBody,
+} from './types';
+
+export const getRequestData = async (
+  request: KibanaRequest<unknown, AlertListRequestParams, AlertListRequestParams>,
+  endpointAppContext: EndpointAppContext
+): Promise<AlertListRequestData> => {
+  const config = await endpointAppContext.config();
+  const reqData = new AlertListRequestData();
+
+  reqData.pageIndex = request.query?.page_index;
+  reqData.pageSize = request.query?.page_size || config.alertResultListDefaultPageSize;
+  reqData.query = request.query?.query || config.alertResultListDefaultQuery;
+  reqData.filters = request.query?.filters || config.alertResultListDefaultFilters;
+  reqData.dateRange = request.query?.date_range || config.alertResultListDefaultDateRange;
+  reqData.sort = request.query?.sort || config.alertResultListDefaultSort;
+  reqData.order = request.query?.order || config.alertResultListDefaultOrder;
+  reqData.searchAfter = request.query?.after;
+  reqData.searchBefore = request.query?.before;
+
+  if (reqData.searchAfter === undefined && reqData.searchBefore === undefined) {
+    // simple pagination
+    if (reqData.pageIndex === undefined) {
+      reqData.pageIndex = config.alertResultListDefaultFirstPageIndex;
+    }
+
+    reqData.fromIndex = reqData.pageIndex * reqData.pageSize;
+  }
+
+  return reqData;
+};
+
+export function buildQuery(reqData: AlertListRequestData): JsonObject {
+  const queries: JsonObject[] = [];
+  let dateRangeFilter: JsonObject = {};
+
+  if (reqData.filters.length > 0) {
+    const filtersQuery = buildQueryFromFilters(reqData.filters, undefined);
+    queries.push(...filtersQuery.filter);
+  }
+
+  if (reqData.query) {
+    queries.push(toElasticsearchQuery(fromKueryExpression(reqData.query)));
+  }
+
+  if (reqData.dateRange) {
+    dateRangeFilter = {
+      range: {
+        ['@timestamp']: {
+          gte: reqData.dateRange.from,
+          lte: reqData.dateRange.to,
+        },
+      },
+    };
+
+    queries.push(dateRangeFilter);
+  }
+
+  // Optimize
+  if (queries.length > 1) {
+    return {
+      bool: {
+        must: queries,
+      },
+    };
+  } else if (queries.length === 1) {
+    return queries[0];
+  }
+
+  return {
+    match_all: {},
+  };
+}
+
+export const buildAlertListESQuery = async (reqData: AlertListRequestData): Promise<JsonObject> => {
+  const DEFAULT_TOTAL_HITS = 10000;
+  let totalHitsMin: number = DEFAULT_TOTAL_HITS;
+
+  // Calculate minimum total hits set to indicate there's a next page
+  // TODO: handle this for search_after?
+  if (reqData.fromIndex !== undefined) {
+    totalHitsMin = Math.max(reqData.fromIndex + reqData.pageSize * 2, DEFAULT_TOTAL_HITS);
+  }
+
+  const reqBody: AlertListRequestBody = {
+    track_total_hits: totalHitsMin,
```

Comment:
nit: If we do need to support page index, another option could be the value count aggregation: https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-metrics-valuecount-aggregation.html
It's an estimate but gets around the track_total_hits limit.

Reply:
@jonathan-buttner Do you think it would impact performance by doing an aggregation every time we search? Not sure which is better, but let's chat about it...